GPflowOpt/acquisition/qei_cl.py

```diff
@@ -0,0 +1,51 @@
+# Copyright 2017 Joachim van der Herten
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from .ei import ExpectedImprovement
+import numpy as np
+
+
+class QEI_CL(ExpectedImprovement):
+    """
+    This class is an implementation of the constant liar heuristic (min case)
+    for using Expected Improvement in the batch case.
+
+    See:
+    Ginsbourger D., Le Riche R., Carraro L. (2010)
+    Kriging Is Well-Suited to Parallelize Optimization.
+    """
+
+    def __init__(self, model, batch_size):
+        super(QEI_CL, self).__init__(model, batch_size=batch_size)
+        self.in_batch = False
+
+    def set_batch(self, *args):
+        assert self.in_batch, 'Set batch must be called within a context'
+
+        X = np.vstack((self.X_,) + args)
+        Y = np.vstack((self.Y_,) + (self.fmin.value,)*len(args))
+        self.set_data(X, Y)
+
+    def __enter__(self):
```

Comment:
Two notes: 
- I dislike the old style of doing contexts and prefer the contextmanager decorator (see https://jeffknupp.com/blog/2016/03/07/python-with-context-managers/ fun with contextlib). 
- I think, with contextlib, this should be moved to Acquisition as the batch context is something which is potentially applicable to any acquisition implementation.

Reply:
what about the following in acquisition:
```python
class SequentialBatch(object):
      def __init__(acquisition, size):
            self.acquisition = acquisition
            self.size = size

     def optimize(optimizer):
         points = []
         def inverse_acquisition(X):
               ....

         for j in range(self.acquisition.batch_size):
             result = optimizer.optimize(inverse_acquisition)
             batch.append(result.x)
             self.acquisition.set_batch(np.concatenate(points))
     return np.concatenate(points)

...

class Acquisition(Parameterized):
....

    def set_batch(self, batch_points):
        pass  # To be implemented in subclasses, e.g. qEI

    @contextmanager
    def batch(size):
          X_orig, Y_orig = self.data
          yield SequentialBatch(self, size)
          self.set_batch(np.empty((0, self.domain.size)))
          self.set_data(X_orig, Y_orig)
```
Then in BayesianOptimizer you could do: 
```python
    if self.batch_size > 1:
        with self.acquisition.batch(self.batch_size) as batch:
            points = batch.optimize(self.optimizer)
       self._update_model_data(points)
```
with some further thought I think you could even see the one point strategy here as a batch with size 1 here and simply do this by default in BayesianOptimizer. What about my proposal for set_batch? It seems intuitive to simply pass in an array which is the current batch. Then if you pass a 0 x dim it signals clearing. This way has some advantages I think
- The interface for adding batch support to an acquisition implementation is limited to implementing set_batch.
- if we'd ever have asynchronous batches, once a point was evaluate you could simply perform `set_batch(2d array \ evaluated point)` and optimize. The great thing is that given this angle, all that would be needed to expand towards such approach is another object which is yielded from the context, e.g. AsyncSequentialBatch or something. Just a motivation for the design in the snippet)

Reply:
One more thing I just realized, with the pseudocode of the previous reply, the logic of selecting the points relays to a different policy object (the SequentialBatch). It would be perfectly possible to add a `TrueBatch` which duplicates the domain and selects the points at once. Swtiching between both approaches would simply require to instantiate BayesianOptimizer with a different object in its constructor. This would be a very simple and clean way to make both approaches co-exist if we merge this branch with the other batch approach.

Reply:
Hey, thanks for the very detailed response! Sorry the delay from my part; the past week was busy.

Having a `batch` `contextmanager` in `Acquisition` assumes that the initialization and finilization for the batch is irrespective of the batch policy. I think this is restrictive, particularly due to the fact that it cannot be "overwritten" by the `SequentialBatch` class. This is why I tried to add the context in the final class itself. Note that, to my understanding, this cannot be done with `contextlib` decorators. On the other hand, I agree that `set_batch` should be included in the base `Acquisition` class as it can be overwritten by other child classes.

If we only plan to use the `contextmanager` to call `optimize`, as you suggest doing in `BayesianOptimizer`:
```
with self.acquisition.batch(self.batch_size) as batch:
    points = batch.optimize(self.optimizer)
```
then there is no need to use the context, at least in `BayesianOptimizer`. The function batch.optimize() should be responsible for initializing/clean up (either with a help of a context internally or by itself).

But I do believe that a `context` is useful. It gives us safe access to penalised/modified acquisition functions via `set_batch`, which would be useful for researching the effect of the batch heuristics used.

I could add a function `optimise` in the batch acquisition class. It would be something like:
```
def optimize(optimizer):
         points = []
         def inverse_acquisition(X):
               ....
         # Use the context here
         with self:
                # Build the batch...
         return points
```
and the `BayesianOptimiser` would simply do:
```
if self.batch_size > 1:
        # No need for context here, it is used inside acquisition.optimize
        points = acquisition.optimize(self.optimizer)
```
Finally, this might be on the pedantic side, but I would call the function `get_suggestion` instead of `optimize`. I think `optimise` is not necessarily descriptive of the process done and it would make me wonder why not directly use `optimizer.optimize`.

Looking forward to hearing your thoughts on this.

Reply:
We have discussed this today and considered some scenario's. I think we won't be able to solve everything in one PR unfortunately, we are proposing the following for now to get started which should be a nice compromise:
(1) do not implement a context for now
(2) Class wise, we'd like to go for the following
<img width="497" alt="image" src="https://user-images.githubusercontent.com/767772/30038737-34538f28-91c8-11e7-9bda-a0a7d8f70949.png">
I favor your idea of naming (i.e. `get_suggestion`) as `optimize` would be confusing with GPflow, and confusing with `optimizer.optimize`.
i.e., the `get_suggestion` in sequential batch looks a lot like the sequential batch above and also contains the operation of the context currently. If a sequential strategy really diverts from the default it can override get_suggestion. Furthermore, the methods get_suggestion relies on go in the interface of the derived class (i.e. in SequentialBatchAcquisition, a set_batch is required, this method is missing from the other classes). In Acquisition `get_suggestion` simply optimizes, and BayesianOptimizer calls acquisition.get_suggestion, it doesn't even need to verify batch size. 


I think this is a good design to start with, its also a plain OO solution to solve this without extra complexity. We also added for instance an ABAcquisition, which could be used to implement for instance https://arxiv.org/pdf/1704.03651.pdf which isn't a true batch approach but it shares some semantics. Unfortunately, this currently isn't compatible with the aggregation mechanics and MCMC components we have in place currently. I'm still in the process of discussing this with @icouckuy but we should be able to resolve this later on it will however involve some reshuffling but right now I think its fully manageable.

What do you think. You could add SequentialBatchAcquisition, I'll do ParallelBatchAcquisition.

Reply:
Yup, I think it's a good plan. I will try to update the PR accordingly for SequentialBatchAcquisition by early next week.

Reply:
Great, to give you an idea: this is where we are heading at the moment:
<img width="707" alt="image" src="https://user-images.githubusercontent.com/767772/30166218-efeec740-93e2-11e7-8d07-b8dd50470d3c.png">
still discussion ongoing though. Ultimately the idea is that Acquisition/SequentialBatchAcquisition become the same thing (set_batch by default doing nothing, for batch support it is sufficient to override). The design of the UML also takes into account the current capabilities of the framework to add and multiply, use MCMC etc. Your code will be easy to fit in so don't worry about this for now.